pmd-core/src/main/java/net/sourceforge/pmd/lang/document/SourceCodePositioner.java

```diff
@@ -0,0 +1,266 @@
+        }
+
+        return globalOffset - lineOffsets[lineIndex] + 1; // 1-based column offsets
+    }
+
+    /**
+     * Finds the offset of a position given (line,column) coordinates.
+     * Returns -1 if the parameters don't identify a caret position in
+     * the wrapped text.
+     *
+     * @param line   Line number (1-based)
+     * @param column Column number (1-based)
+     *
+     * @return Text offset (zero-based), or -1
+     */
+    public int offsetFromLineColumn(final int line, final int column) {
+        if (!isValidLine(line)) {
+            if (line == lineOffsets.length && column == 1) {
+                return sourceCodeLength;
+            }
+            return -1;
+        }
+
+        final int lineIdx = line - 1;
+        int bound = offsetOfEndOfLine(line);
+        int off = lineOffsets[lineIdx] + column - 1;
+        return off > bound ? -1 : off;
+    }
+
+    /**
+     * Returns the offset of the end of the given line. This is the caret
+     * position that follows the last character on the line (which includes
+     * the line terminator if any). This is the caret position at the
+     * start of the next line, except if the line is the last in the document.
+     *
+     * @param line Line number (1-based)
+     *
+     * @return Text offset
+     *
+     * @throws IndexOutOfBoundsException If the line is invalid
+     */
+    public int offsetOfEndOfLine(final int line) {
+        if (!isValidLine(line)) {
+            throw new IndexOutOfBoundsException(
+                line + " is not a valid line number, expected at most " + lineOffsets.length);
+        }
+
+        return lineOffsets[line];
+    }
+
+    boolean isValidLine(int line) {
+        return line >= 1 && line <= getLastLine();
+    }
+
+    /**
+     * Returns the number of lines, which is also the ordinal of the
+     * last line.
+     */
+    public int getLastLine() {
+        return lineOffsets.length - 1;
+    }
+
+    public int getNumLines() {
+        return getLastLine();
+    }
+
+    /**
+     * Returns the last column number of the last line in the document.
+     */
+    public int getLastLineColumn() {
+        return getLastColumnOfLine(getLastLine());
+    }
+
+    private int getLastColumnOfLine(int line) {
+        if (line == 0) {
+            return 1 + lineOffsets[line];
+        } else {
+            return 1 + lineOffsets[line] - lineOffsets[line - 1];
+        }
+    }
+
+    /**
+     * Builds a new positioner for the given char sequence.
+     * The char sequence should not change state (eg a {@link StringBuilder})
+     * after construction, otherwise this positioner becomes unreliable.
+     *
+     * @param charSeq Text to wrap
+     */
+    public static SourceCodePositioner create(CharSequence charSeq) {
+        final int len = charSeq.length();
+        Builder builder = new Builder();
+
+        int off = 0;
+        while (off < len) {
+            char c = charSeq.charAt(off);
+            if (c == '\n') {
```

Comment:
This code handles `\r\n` correctly, as both `\r\n` and `\n` end with the same char. I realise it doesn't handle `\r`, I only very recently added support for that in TextFileContent. The charsequence is not expected to be normalized here so that code needs to be updated

Reply:
I think in fact, the charsequence is expected to be normalized here... I changed the visibility to package private